File: app/schemas/flight.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from app.models.flight import FlightStatus


class FlightBase(BaseModel):
    flight_number: str
    origin_id: int
    destination_id: int
    airplane_id: int
    scheduled_departure: datetime
    scheduled_arrival: datetime
    gate: Optional[str] = None
    terminal: Optional[str] = None
    status: Optional[FlightStatus] = FlightStatus.SCHEDULED


class FlightCreate(FlightBase):
    pass


class FlightUpdate(BaseModel):
    origin_id: Optional[int] = None
    destination_id: Optional[int] = None
    airplane_id: Optional[int] = None
    scheduled_departure: Optional[datetime] = None
    scheduled_arrival: Optional[datetime] = None
    gate: Optional[str] = None
    terminal: Optional[str] = None
    status: Optional[FlightStatus] = None


class FlightSearch(BaseModel):
    origin: Optional[str] = None  # Airport code
    destination: Optional[str] = None  # Airport code
    date: Optional[datetime] = None


class FlightResponse(FlightBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class FlightDetailResponse(FlightResponse):
    origin_airport: dict
    destination_airport: dict
    airplane: dict
    available_seats: int
    total_seats: int

File: frontend/lib/models/flight.dart
class Airport {
  final int id;
  final String code;
  final String name;
  final String city;
  final String country;

  Airport({
    required this.id,
    required this.code,
    required this.name,
    required this.city,
    required this.country,
  });

  factory Airport.fromJson(Map<String, dynamic> json) {
    return Airport(
      id: json['id'],
      code: json['code'],
      name: json['name'],
      city: json['city'],
      country: json['country'],
    );
  }
}

class Flight {
  final int id;
  final String flightNumber;
  final String scheduledDeparture;
  final String scheduledArrival;
  final String status;
  final Airport originAirport;
  final Airport destinationAirport;

  Flight({
    required this.id,
    required this.flightNumber,
    required this.scheduledDeparture,
    required this.scheduledArrival,
    required this.status,
    required this.originAirport,
    required this.destinationAirport,
  });

  factory Flight.fromJson(Map<String, dynamic> json) {
    return Flight(
      id: json['id'],
      flightNumber: json['flight_number'],
      scheduledDeparture: json['scheduled_departure'],
      scheduledArrival: json['scheduled_arrival'],
      status: json['status'],
      originAirport: Airport.fromJson(json['origin_airport']),
      destinationAirport: Airport.fromJson(json['destination_airport']),
    );
  }

  DateTime get departureTime => DateTime.parse(scheduledDeparture);
  DateTime get arrivalTime => DateTime.parse(scheduledArrival);
}
